File: minimum_falling_path.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        arr=matrix
        nrows=len(matrix)
        ncols = len(matrix[0])
        dp = [[None for i in range(ncols)] for j in range(nrows)]
        dp[0] = matrix[0][:]
        for i in range(1, nrows):
            for j in range(ncols):
                if j == 0:
                    dp[i][j] = min(dp[i - 1][j], dp[i - 1][j + 1]) + matrix[i][j]
                elif j == len(matrix[0]) - 1:
                    dp[i][j] = min(dp[i - 1][j], dp[i - 1][j - 1]) + matrix[i][j]
                else:
                    dp[i][j] = min(dp[i - 1][j], dp[i - 1][j + 1], dp[i - 1][j - 1]) + matrix[i][j]
        return min(dp[-1])

File: minimumFallingPath.cpp
//Time complexity: O(n*n)
//Space complexity: O(1)

//Approach: idea is to start with a second last row and keep chosing minimum value form one row below it. then onwards keep going until first row.
//in the end calculate minimum between all the column values.

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int m=matrix.size();
        int n=matrix[0].size();

        for(int i=n-2; i>=0; --i)
        {
            for(int j=0;j<n;j++)
            {
                if(j==0)
                {
                    matrix[i][j]=matrix[i][j] + min(matrix[i+1][j], matrix[i+1][j+1]);
                }
                else if(j==n-1)
                {
                    matrix[i][j]=matrix[i][j]+ min(matrix[i+1][j],matrix[i+1][j-1]);
                }
                else
                {
                    matrix[i][j]=matrix[i][j]+min(min(matrix[i+1][j],matrix[i+1][j+1]), matrix[i+1][j-1]);
                }
            }
        }
        int min_=INT_MAX;
        for(int i=0;i<n;++i)
        {
            min_=min(min_, matrix[0][i]);
        }
        return min_;
    }
};